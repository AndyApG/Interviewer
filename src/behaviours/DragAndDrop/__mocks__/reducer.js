/* eslint-env jest */

const upsertTarget = jest.fn(() => ({ type: null }));
const renameTarget = jest.fn(() => ({ type: null }));
const removeTarget = jest.fn(() => ({ type: null }));

const dragStart = jest.fn(() => ({ type: null }));
const dragMove = jest.fn(() => ({ type: null }));
const dragEnd = jest.fn(() => ({ type: null }));

const actionCreators = {
  upsertTarget,
  renameTarget,
  removeTarget,
  dragStart,
  dragMove,
  dragEnd,
};

const reducer = jest.fn();

export {
  actionCreators,
  reducer,
};

export default reducer;
