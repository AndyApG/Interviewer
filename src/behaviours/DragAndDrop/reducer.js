import { filter, reject, omit, isEmpty, thru } from 'lodash';

const UPDATE_TARGET = Symbol('DRAG_AND_DROP/UPDATE_TARGET');
const RENAME_TARGET = Symbol('DRAG_AND_DROP/RENAME_TARGET');
const REMOVE_TARGET = Symbol('DRAG_AND_DROP/REMOVE_TARGET');
const DRAG_START = Symbol('DRAG_AND_DROP/DRAG_START');
const DRAG_MOVE = Symbol('DRAG_AND_DROP/DRAG_MOVE');
const DRAG_END = Symbol('DRAG_AND_DROP/DRAG_END');

const initialState = {
  targets: [],
  source: null,
};

const defaultSource = {
  meta: {},
};

// Since accepts() is a weak link and can throw errors if not carefully written.
const willAccept = (accepts, source) => {
  try {
    return accepts({
      ...defaultSource,
      ...source,
    });
  } catch (e) {
    console.log('Error in accept() function', e, source); // eslint-disable-line no-console
    return false;
  }
};

const markHitTarget = ({ target, source }) => {
  if (!source) { return { ...target, isOver: false, willAccept: false }; }

  const isOver = (
    source.x > target.x &&
    source.x < target.x + target.width &&
    source.y > target.y &&
    source.y < target.y + target.height
  );

  return { ...target, isOver, willAccept: willAccept(target.accepts, source) };
};

const markHitTargets = ({ targets, source }) =>
  targets.map(target => markHitTarget({ target, source }));

const markHitSource = ({ targets, source }) =>
  thru(source, (s) => {
    if (isEmpty(s)) { return s; }

    return {
      ...s,
      isOver: filter(targets, 'isOver').length > 0,
    };
  });

const markHitAll = ({ targets, source, ...rest }) => {
  const targetsWithHits = markHitTargets({ targets, source });
  const sourceWithHits = markHitSource({ targets: targetsWithHits, source });

  return {
    ...rest,
    targets: targetsWithHits,
    source: sourceWithHits,
  };
};

const resetHits = ({ targets, ...rest }) => ({
  targets: targets.map(target => omit(target, ['isOver', 'willAccept'])),
  ...rest,
});

const triggerDrag = (state, source) => {
  const hits = markHitAll({
    ...state,
    source: {
      ...state.source,
      ...source,
    },
  });

  filter(hits.targets, { isOver: true, willAccept: true })
    .forEach((target) => {
      target.onDrag(hits.source);
    });
};

const triggerDrop = (state, source) => {
  const hits = markHitAll({
    ...state,
    source: {
      ...state.source,
      ...source,
    },
  });

  filter(hits.targets, { isOver: true, willAccept: true })
    .forEach((target) => {
      target.onDrop(hits.source);
    });
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case UPDATE_TARGET: {
      const targets = [
        ...reject(state.targets, ['id', action.target.id]),
        markHitTarget({ target: action.target, source: state.source }),
      ];

      const source = markHitSource({
        targets,
        source: state.source,
      });

      return {
        ...state,
        targets,
        source,
      };
    }
    case RENAME_TARGET:
      return {
        ...state,
        targets: state.targets.map((target) => {
          if (action.from !== target.id) { return target; }

          return {
            ...target,
            id: action.to,
          };
        }),
      };
    case REMOVE_TARGET: {
      const targets = reject(state.targets, ['id', action.id]);
      const source = markHitSource({ targets, source: state.source });
      return {
        ...state,
        targets,
        source,
      };
    }
    case DRAG_START: {
      return markHitAll({
        ...state,
        source: action.source,
      });
    }
    case DRAG_MOVE:
      if (state.source === null) { return state; }

      return markHitAll({
        ...state,
        source: {
          ...state.source,
          ...action.source,
        },
      });
    case DRAG_END:
      return resetHits({
        ...state,
        source: null,
      });
    default:
      return state;
  }
};

function updateTarget(data) {
  return {
    type: UPDATE_TARGET,
    target: data,
  };
}

function renameTarget({ from, to }) {
  return {
    type: RENAME_TARGET,
    from,
    to,
  };
}

function removeTarget(id) {
  return {
    type: REMOVE_TARGET,
    id,
  };
}

function dragStart(data) {
  return {
    type: DRAG_START,
    source: data,
  };
}

function dragMove(data) {
  return (dispatch, getState) => {
    triggerDrag(getState(), data);

    dispatch({
      type: DRAG_MOVE,
      source: data,
    });
  };
}

function dragEnd(data) {
  return (dispatch, getState) => {
    triggerDrop(getState(), data);

    dispatch({
      type: DRAG_END,
    });
  };
}


const actionCreators = {
  updateTarget,
  renameTarget,
  removeTarget,
  dragStart,
  dragMove,
  dragEnd,
};

const actionTypes = {
  UPDATE_TARGET,
  RENAME_TARGET,
  REMOVE_TARGET,
  DRAG_START,
  DRAG_MOVE,
  DRAG_END,
};

export {
  actionCreators,
  actionTypes,
  reducer,
};

export default reducer;
