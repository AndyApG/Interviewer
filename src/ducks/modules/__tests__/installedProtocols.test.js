/* eslint-env jest */

import reducer from '../installedProtocols';

const initialState = {};

describe('protocols reducer', () => {
  it('should return the initial state', () => {
    expect(
      reducer(undefined, {}),
    ).toEqual(initialState);
  });

  // describe('SET_PROTOCOL', () => {
  //   it('adds a new protocol', () => {
  //     const setProtocolAction = { type: ProtocolActionTypes.SET_PROTOCOL, installedProtocols: { name: 'new' } };
  //     expect(reducer(initialState, setProtocolAction)).toHaveLength(initialState.length + 1);
  //   });

  //   it('does not add an existing protocol', () => {
  //     const setProtocolAction = { type: ProtocolActionTypes.SET_PROTOCOL, installedProtocols: { name: 'new' } };
  //     const newState = reducer(initialState, setProtocolAction);
  //     expect(reducer(newState, setProtocolAction)).toHaveLength(newState.length);
  //   });
  // });
});
