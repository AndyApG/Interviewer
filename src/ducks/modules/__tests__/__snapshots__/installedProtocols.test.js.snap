// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`protocols reducer DELETE_PROTOCOL Has existing sessions warns user about existing sessions 1`] = `
Array [
  Array [
    Object {
      "canCancel": false,
      "message": "This protocol has in-progress sessions that have not yet been exported. Please export or delete these sessions before attempting to delete the protocol.",
      "title": "Existing sessions",
      "type": "Warning",
    },
  ],
]
`;

exports[`protocols reducer DELETE_PROTOCOL confirms and then deletes protocol 1`] = `
Array [
  Array [
    Object {
      "protocolUID": "1234",
      "type": "INSTALLED_PROTOCOLS/DELETE_PROTOCOL",
    },
  ],
]
`;

exports[`protocols reducer DELETE_PROTOCOL does not confirm and does not delete protocol 1`] = `Array []`;
